lighthouse-core/gather/gather-runner.js

```diff
@@ -108,14 +110,21 @@ class GatherRunner {
    * Navigates to about:blank and calls beforePass() on gatherers before tracing
    * has started and before navigation to the target page.
    * @param {!Object} options
+   * @param {!Object<!Array<!Promise<*>>} gathererResults
    * @return {!Promise}
    */
-  static beforePass(options) {
+  static beforePass(options, gathererResults) {
     const pass = GatherRunner.loadBlank(options.driver);
 
     return options.config.gatherers.reduce((chain, gatherer) => {
       return chain.then(_ => {
-        return gatherer.beforePass(options);
+        const artifactPromise = Promise.resolve().then(_ => gatherer.beforePass(options));
+        gathererResults[gatherer.name] = [artifactPromise];
+        return artifactPromise.catch(err => {
```

Comment:
thoughts on just extracting this piece from the 3 *pass functions? it's the ugliest part and wouldn't need any crazy parameter list. would also DRY up recovery in case we make it more complicated or something

`return this._recoverOrThrow(artifactPromise)`